made the bee a basic insect

Ben10Omintrix opened the pull request:
## About The Pull Request
the bee now a baisc insect he will now go to  find his home and he will go and pollinated the plants and helped the queen make children by polliniting the plants and he will. the queen will leve the hive more rarely than the normal bees so she can stay in the hive to make kids

## Why It's Good For The Game
the bee now is a basic insect so it means he have a better ai

## Changelog
:cl:
balance: the bee now can fly over the machines so its easy for him to go to the hydroponics machine
fix: player bees now will not be stuck inside the hive if he entered it, they can now leave it
fix: fixed a har deleted when the hive is deleted all the bees still have a refence to the hive now its fixed
fix: now when a player interacted with the bee hive the bees will now leave the hive to defend the hive (it was glitched)
refactor: the bees now are a basic insect.
/:cl:

Ben10Omintrix committed:
tgstation pulle request

Ben10Omintrix committed:
Update _bee.dm

san7890 commented on code/modules/research/experimentor.dm:
```diff
@@ -648,7 +648,7 @@
 		/mob/living/simple_animal/crab,
 		/mob/living/simple_animal/hostile/bear,
-		/mob/living/simple_animal/hostile/bee,
+		/mob/living/basic/bee,
```
alphabetize this list.

san7890 commented on code/modules/events/vent_clog.dm:
```diff
@@ -221,7 +221,7 @@
 	var/static/list/mob_list = list(
 		/mob/living/basic/mouse/rat,
-		/mob/living/simple_animal/hostile/bee,
+		/mob/living/basic/bee,
```
alphabetize

san7890 commented on code/modules/events/vent_clog.dm:
```diff
@@ -256,7 +256,7 @@
 	var/static/list/mob_list = list(
 		/mob/living/basic/carp,
-		/mob/living/simple_animal/hostile/bee/toxin,
+		/mob/living/basic/bee/toxin,
```
alphabetize

san7890 commented on code/modules/hydroponics/beekeeping/beebox.dm:
```diff
@@ -255,9 +254,9 @@
-	for(var/mob/living/simple_animal/hostile/bee/B in bees)
-		if(B.loc == src)
-			B.forceMove(drop_location())
+	for(var/mob/living/basic/worker in bees)
```
this spawns the abstract type `/mob/living/basic`, not correct.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+#define BEE_TRAY_RECENT_VISIT 200 //How long in deciseconds until a tray can be visited by a bee again
```
```suggestion
#define BEE_TRAY_RECENT_VISIT 20 SECONDS //! How long until a tray can be visited by a bee again
```

no need for that un-necessary portion of the comment, we can nest defines like this. the `//!` is so it can be dmdocced, but would be nice if you could do a more traditional format for all of these.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+	habitable_atmos = list("min_oxy" = 0, "max_oxy" = 0, "min_plas" = 0, "max_plas" = 0, "min_co2" = 0, "max_co2" = 0, "min_n2" = 0, "max_n2" = 0)
+	basic_mob_flags = DEL_ON_DEATH
+	ai_controller = /datum/ai_controller/basic_controller/bee
+	///he he he, beegent
```
it's funny but i think you should be more descriptive

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+	var/obj/structure/beebox/beehome = null
+	///our icon base
+	var/icon_base = "bee"
+	var/isqueen = FALSE
```
sounds self-explanatory but i'd like if you could DMdoc for statistical purposes. would also prefer if this could be re-named `is_queen` to keep in the `snake_case` standard of this codebase.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+	beegent = null
+	return ..()
+
+/mob/living/basic/bee/UnarmedAttack(atom/target, proximity_flag, list/modifiers)
```
if we're going to override the whole UnarmedAttack proc I think we should use the signal instead. I know there was a bit of a discussion between Melbert and I in another PR but I think if we're going to be killing the entire proc anyhow (and not invoking parent at the start) there shouldn't really be a reason why we shouldn't do this on a signal... any concerns about this though? I don't know if the comsig for attack allows for return types (e.g. some sort of `COMPONENT_CANCEL_SIGNAL` thing)

MrMelbert replied:
I don't think there's any immediately wrong with overriding `UnarmedAttack` since it's the highest level part of the click chain you can get on the mob without going directly into Click

(However you should check proximity flag, it's only set by telekinesis but you never know)

That being said apparently basic mobs do have `/melee_attack` which is called from `UnarmedAttack` which might be more appropriate to override-while-conditionally-calling-parent

san7890 replied:
i would be amenable to either `melee_attack()` or a signal handler for this then, i just don't like completely glonking the click chain which is going to always be subject to a bunch of bullshit in the past, present, and future.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+	beehome.bees += src
+	if(isqueen)
+		beehome.queen_bee = src
+	return
```
```suggestion
```

no need for this return.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+/mob/living/basic/bee/examine(mob/user)
+	. = ..()
+
+	if(!beehome)
```
```suggestion
	if(isnull(beehome))
```

faster, more explicit.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+	if(beehome)
+		beehome.bees -= src
+		beehome = null
+	if((flags_1 & HOLOGRAM_1))
```
```suggestion
	if(flags_1 & HOLOGRAM_1)
```

redundant

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+	beegent = null
+	return ..()
+
+/mob/living/basic/bee/death(gibbed)
```
you've defined `death()` twice. the other one is on line 78.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+
+/mob/living/basic/bee/proc/apply_chemicals(mob/living/basic/attacker, atom/target)
+	SIGNAL_HANDLER
+	if(!beegent || !isliving(target))
```
```suggestion
	if(isnull(beegent) || !isliving(target))
```

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+/mob/living/basic/bee/proc/reagent_incompatible(mob/living/basic/bee/ruler)
+	if(!ruler)
+		return FALSE
+	if(ruler.beegent?.type != beegent?.type)
```
can we do an early return if we have no beegent? just `isnull(beegent)`, given how often we access it.

Ben10Omintrix replied:
the bee can entered the hive even if he dont have a beegent

san7890 replied:
This is `reagent_incompatible()` though? what's that got to do with this?

san7890 replied:
oh you changed it in the code nvm then

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+/mob/living/basic/bee/proc/generate_bee_visuals()
+	cut_overlays()
+
+	var/col = BEE_DEFAULT_COLOUR
```
```suggestion
	var/bee_color = BEE_DEFAULT_COLOUR
```

let's use our words here.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+	if(beehome)
+		beehome.bee_resources = min(beehome.bee_resources + growth, 100)
+
+/mob/living/basic/bee/proc/assign_reagent(datum/reagent/toxin)
+	if(istype(toxin))
```
change this to an early return.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+/mob/living/basic/bee/queen/will_escape_storage()
+	return FALSE
+
+/mob/living/basic/bee/toxin
```
should we add a `desc` or something here? your call.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+	desc = "No wonder the bees are dying out, you monster."
+	icon = 'icons/mob/simple/bees.dmi'
+	icon_state = "bee_item"
+	var/datum/reagent/beegent
+	var/bee_type = /mob/living/basic/bee
```
dmdoc.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+		if(!QDELETED(src))
+			qdel(src)
+
+/obj/item/queen_bee/attackby(obj/item/syringe, mob/user, params)
```
again, since we override all of `attackby()`, i feel like signals would be optimal?

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+
+/obj/item/queen_bee/Exited(atom/movable/gone, direction)
+	. = ..()
+	if(gone == queen)
```
make this an early return?

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+	. = ..()
+	AddElement(/datum/element/swabable, CELL_LINE_TABLE_QUEEN_BEE, CELL_VIRUS_TABLE_GENERIC_MOB, 1, 5)
+
+/obj/item/queen_bee/Destroy()
+	QDEL_NULL(queen)
```
damn really?

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/bee_ai.dm:
```diff
@@ -0,0 +1,175 @@
+	finding_behavior = /datum/ai_behavior/find_hunt_target/pollinate
+	hunt_targets = list(/obj/machinery/hydroponics)
+	hunt_range = 7
+	//hunt_chance = 10
```
no commented out code like this, either add it in fully or delete the line.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/bee_ai.dm:
```diff
@@ -0,0 +1,175 @@
+/datum/ai_behavior/find_hunt_target/pollinate
+
+/datum/ai_behavior/find_hunt_target/pollinate/valid_dinner(mob/living/source, obj/machinery/hydroponics/dinner, radius)
+	if(!dinner.myseed)
```
`isnull()`?

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/bee_ai.dm:
```diff
@@ -0,0 +1,175 @@
+	var/obj/structure/beebox/current_home = controller.blackboard[target_key]
+	var/mob/living/bee_pawn = controller.pawn
+
+	bee_pawn.UnarmedAttack(current_home)
```
Unarmed attack? I think we should split that behavior out of UnarmedAttack out into a new board and then store it as a `/datum/callback` in the blackboard or something that we can execute since we definitely know that this is going to execute the enter hive behavior. Going to have really odd consequences if this behavior is drag and dropped anywhere else. Would allow it to be infinitely more extensible to more mobs that might need to pop in and out of hives without being reliant on unarmed attack.

Ben10Omintrix replied:
i putted him in the unarmedatack so if the bee is sentient he can interacted with the objects the ai can interacted with

san7890 replied:
Forgot to elaborate, this would be done in conjunction with the other suggestion that the earlier behavior should be split out into a signal (or at least invoked via `melee_attack()`) so that both the player bee and the AI bee can use it by either of those interactions.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/bee_ai.dm:
```diff
@@ -0,0 +1,175 @@
+		finish_action(controller, FALSE, target_key)
+		return
+
+	bee_pawn.UnarmedAttack(potential_home) //interact with the house to inhabit
```
ditto on splitting it out of UnarmedAttack and into a callback proc.

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/bee_ai.dm:
```diff
@@ -0,0 +1,175 @@
+
+/datum/targetting_datum/basic/bee
+
+/datum/targetting_datum/basic/bee/can_attack(mob/living/owner, atom/target)
+	if(!isliving(target))
+		return FALSE
+	. = ..()
```
isn't there a `pre_attack` check proc for this? why do we invoke parent in the middle?

Ben10Omintrix replied:
i cant find this proc

san7890 replied:
i could've sworn there was something we check in the parent of `can_attack`. sigh

san7890 replied:
or is `can_attack()` the pre-check code itself? ugh

san7890 reviewed (changes requested):
Mostly nits, although a few serious code issues were noted.

san7890 commented:
lol wait why didn't github show me this was drafted? sorry, still should address a good chunk of this stuff though.

Ben10Omintrix committed:
tgstation pulled request

Ben10Omintrix committed:
Update _bee.dm

Ben10Omintrix committed:
Update _bee.dm

Ben10Omintrix committed:
Update _bee.dm

MrMelbert commented on code/modules/hydroponics/beekeeping/beebox.dm:
```diff
@@ -47,9 +44,13 @@
-	bees.Cut()
-	honeycombs.Cut()
 	queen_bee = null
+	for(var/mob/living/basic/bee/worker in bees)
+		bees -= worker
+		worker.beehome = null
+		worker.forceMove(get_turf(src))
```
`Destroy` should not have side effects, any remaining workers should be qdeleted (which is what is currently being done, any bees in the contents of the box are qdel'd)

You can however also implement `/deconstruct` and put this code here, so if someone bashes the beebox to bits, the bees will flow out instead.

MrMelbert commented on code/modules/hydroponics/beekeeping/beebox.dm:
```diff
@@ -198,16 +199,14 @@
-			var/mob/living/simple_animal/hostile/bee/B = b
-			if(B.isqueen)
+		var/bees_attack = FALSE
+		for(var/mob/living/basic/bee/worker in bees) //everyone who's ever lived here now instantly hates you, suck it assistant!
```
`as anything`, if the bees list is only bees

MrMelbert commented on code/modules/hydroponics/beekeeping/beebox.dm:
```diff
@@ -255,9 +254,9 @@
-	for(var/mob/living/simple_animal/hostile/bee/B in bees)
-		if(B.loc == src)
-			B.forceMove(drop_location())
+	for(var/mob/living/basic/bee/worker in bees)
```
`as anything`

MrMelbert commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,338 @@
+		beehome = null
+	if(flags_1 & HOLOGRAM_1)
+		return ..()
+	new /obj/item/trash/bee(loc, src)
```
Small thing but it shouldn't drop bee trash if `gibbed` is `TRUE`

san7890 commented:
i think you forgot to push the commits that addressed melbert's reviews, i unmarked them

Ben10Omintrix committed:
tgstation pulled request

Ben10Omintrix committed:
tgstation pull request

Ben10Omintrix commented:
> i think you forgot to push the commits that addressed melbert's reviews, i unresolved them for the time being. I also addressed your comments and elaborated on some stuff I wrote.

i am finished the changes u want and i tested him and its working. u can check the code again?

Ben10Omintrix committed:
Update _bee.dm

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -0,0 +1,339 @@
+
+/obj/item/trash/bee/Initialize(mapload, mob/living/basic/bee/dead_bee)
+	. = ..()
+	if(dead_bee)
+		pixel_x = dead_bee.pixel_x
+		pixel_y = dead_bee.pixel_y
+		bee_type = dead_bee.type
+		if(dead_bee.beegent)
+			beegent = dead_bee.beegent
+			reagents.add_reagent(beegent.type, 5)
+		update_appearance()
+	AddComponent(/datum/component/edible, list(/datum/reagent/consumable/nutriment/vitamin = 5), null, BEE_FOODGROUPS, 10, 0, list("bee"), null, 10)
+	AddElement(/datum/element/swabable, CELL_LINE_TABLE_QUEEN_BEE, CELL_VIRUS_TABLE_GENERIC_MOB, 1, 5)
+	RegisterSignal(src, COMSIG_ATOM_ON_LAZARUS_INJECTOR, PROC_REF(use_lazarus))
```
reorganize this entire proc to be an early return (so just do the whole addcomponent/element/signal stuff first, then early return on an `if(isnull(dead_bee))`

san7890 reviewed (changes requested):
my last thoughts on the matter

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/bee_ai.dm:
```diff
@@ -0,0 +1,181 @@
+
+
+/datum/ai_controller/basic_controller/bee
```
i also think this whole implementation would look a lot cleaner if you split this file up into multiple files i.e. `bee_subtrees.dm`, `bee_behavior.dm`, etc. seeing as we already have a full folder for bee stuff already.

Ben10Omintrix committed:
tgstation pull request

Ben10Omintrix committed:
Update _bee.dm

Ben10Omintrix committed:
tgstation pul lrequest

Ben10Omintrix committed:
Update _bee.dm

Ben10Omintrix committed:
Update _bee.dm

san7890 commented on code/modules/mob/living/basic/farm_animals/bee/_bee.dm:
```diff
@@ -130,15 +129,6 @@
 	if(is_queen)
 		beehome.queen_bee = src
 
-/mob/living/basic/bee/proc/apply_chemicals(mob/living/basic/attacker, atom/target)
```
why was this removed, is it done elsewhere?

Ben10Omintrix replied:
yes. he get a venemous element in assign_reagent() so this is dont need now
